text/0045-docker-proxy.md

```diff
@@ -0,0 +1,167 @@
+- Feature Name: docker_proxy
+- Start Date: 2019-07-01
+- RFC PR: https://github.com/hyperledger/sawtooth-rfcs/pull/45
+- Sawtooth Issue: N/A
+
+# Summary
+[summary]: #summary
+
+Sawtooth has containerized build and test processes for core and all
+components. Running Docker containers behind a network proxy requires
+special handling so that the containers can find external routes to download
+packages and build dependencies, for example. There are a number of
+ways to configure these environment for proxy support. It is desirable that we
+adopt a single consistent approach across the large body of Docker and
+Docker Compose files in the Sawtooth project.
+
+This RFC defines the canonical approach for supporting proxies for
+Sawtooth containers using Docker 18.x and Docker Compose 1.24.x with
+Ubuntu 16.04 and 18.04 and serves as guidelines to resolve these issues.
+
+# Motivation
+[motivation]: #motivation
+
+Sawtooth repositories make extensive use of Docker containers to build and
+test the software as well as provide turn key examples. Those containers
+typically need to fetch external packages or make other outbound network
+calls. The containers need special configuration to be executed behind
+network proxies. There are multiple ways to provide that proxy configuration
+and it is desirable that the Sawtooth repositories standardize on a single
+approach. The challenge is further complicated by different abilities and
+constraints of different versions of Docker and Ubuntu.
+
+Starting with Ubuntu 18 (Bionic), `gpg` , invoked through `apt-key adv` ,
+now ignores `http_proxy` and `https_proxy` environment variables, and ignores
+the command line options
+`apt-key adv --keyserver-options http-proxy=<proxy server url>`
```

Comment:
Ubuntu 18 doesn't ignore command line options, however it would be too much of a code in every Dockerfile if we choose this way.

Reply:
The current version of `gpg` does ignore the `http-proxy` and `https-proxy` command line options. I found this out from testing on Bionic and confirmed it with multiple documents on the web. This document, for example, says:

```http-proxy=value
...
These options have no more function since GnuPG 2.1```

https://www.gnupg.org/documentation/manuals/gnupg/GPG-Configuration-Options.html

Reply:
Ah! I am not sure what version of gpg gets installed. These CLI worked for me. Thanks for pointing it out.

Reply:
Please include the gnupg link above in the RFC.

Reply:
Added a link.